# Procesado de datos (preprocessing.py)

En esta libreta voy a probar los métodos de preprocessing.py, para visualizar cada paso desde que se leen los datos hasta que se separan y escalan, dejándolos listos para usarlos.

In [5]:
import sys
sys.path.append("..")

# importo los metodos de src/preprocessing.py
from src.preprocessing import cargar_datos, separar_escalar_datos

# 1. Carga y visualización de datos

In [6]:
df = cargar_datos("../data/creditcard.csv")
df.head()

,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0.0,-1.359807,-0.072781,2.536347,1.378155,-0.338321,0.462388,0.239599,0.098698,0.363787,...,-0.018307,0.277838,-0.110474,0.066928,0.128539,-0.189115,0.133558,-0.021053,149.62,0
1,0.0,1.191857,0.266151,0.166480,0.448154,0.060018,-0.082361,-0.078803,0.085102,-0.255425,...,-0.225775,-0.638672,0.101288,-0.339846,0.167170,0.125895,-0.008983,0.014724,2.69,0
2,1.0,-1.358354,-1.340163,1.773209,0.379780,-0.503198,1.800499,0.791461,0.247676,-1.514654,...,0.247998,0.771679,0.909412,-0.689281,-0.327642,-0.139097,-0.055353,-0.059752,378.66,0
3,1.0,-0.966272,-0.185226,1.792993,-0.863291,-0.010309,1.247203,0.237609,0.377436,-1.387024,...,-0.108300,0.005274,-0.190321,-1.175575,0.647376,-0.221929,0.062723,0.061458,123.50,0
4,2.0,-1.158233,0.877737,1.548718,0.403034,-0.407193,0.095921,0.592941,-0.270533,0.817739,...,-0.009431,0.798278,-0.137458,0.141267,-0.206010,0.502292,0.219422,0.215153,69.99,0


Las características principales son **Time** y **Amount**. Las demás características V1-V28 son características que fueron transformadas mediante la técnica PCA. 
Protege información sensible como números de tarjeta, nombre, ubicaciones, hora, país... y las transforma en 28 características no correlacionadas entre sí.

# 2. Separación y escalado

El siguiente paso es separar el dataset en train y test, y escalarlo. Todo eso lo hace la segunda función de preprocessing.py, **separar_escalar_datos**.

In [7]:
X_train, X_test, y_train, y_test = separar_escalar_datos(df)

In [9]:
# Ahora voy a analizar los datos separados
print(f"Dimensiones de X_train: {X_train.shape}")
print(f"Dimensiones de y_train: {y_train.shape}")
print(f"Dimensiones de X_test: {X_test.shape}")
print(f"Dimensiones de y_test: {y_test.shape}")

Dimensiones de X_train: (227845, 30)
Dimensiones de y_train: (227845,)
Dimensiones de X_test: (56962, 30)
Dimensiones de y_test: (56962,)


La función está preparada para separar según: 80% para train - 20% para test. Se puede ver que train cuenta con 227845 ejemplares, y test con 56962.

In [11]:
print(X_train[:5])

            Time        V1        V2        V3        V4        V5        V6  \
265518  0.905774  1.946747 -0.752526 -1.355130 -0.661630  1.502822  4.024933   
180305  0.465984  2.035149 -0.048880 -3.058693  0.247945  2.943487  3.298697   
42664  -0.512286 -0.991920  0.603193  0.711976 -0.992425 -0.825838  1.956261   
198723  0.561678  2.285718 -1.500239 -0.747565 -1.668119 -1.394143 -0.350339   
82325  -0.298886 -0.448747 -1.011440  0.115903 -3.454854  0.715771 -0.147490   

              V7        V8        V9  ...       V20       V21       V22  \
265518 -1.479661  1.139880  1.406819  ... -0.134435  0.076197  0.297537   
180305 -0.002192  0.674782  0.045826  ... -0.227279  0.038628  0.228197   
42664  -2.212603 -5.037523  0.000772  ...  1.280856 -2.798352  0.109526   
198723 -1.427984  0.010010 -1.118447  ... -0.490642 -0.139670  0.077013   
82325   0.504347 -0.113817 -0.044782  ... -0.275297 -0.243245 -0.173298   

             V23       V24       V25       V26       V27       V28  

En estas primeras 5 líneas de X_train se pueden ver dos cosas:
- Ya no tiene la columna Class, que está almacenada en y_train.
- Las columnas time y amount están normalizadas, lo que permite que el resultado del modelo final sea más preciso.

# 3. Comprobar que existen fraudes
Voy a comprobar que existen fraudes tanto en y_train como en y_test. La proporción debe ser muy baja.

In [ ]:
y_train.value_counts(normalize=True)

Class
0    0.998271
1    0.001729
Name: proportion, dtype: float64

: 